### PR TITLE
feat: bind DID call to sender & split did origin in subject and submitter

### PR DESCRIPTION
**Pull request by weichweich**
* Split the origin in submitter and subject
* make sure that the did call can only be send by a specific sender

## new types:

https://github.com/KILTprotocol/type-definitions/pull/21

## Checklist:

- [x] I have verified that the code works
  - [x] No panics! (checked arithmetic ops, no indexing `array[3]` use `get(3)`, ...)
- [x] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [x] ~This PR does not introduce new custom types~
  - [x] If not, I have opened a companion PR with the type changes in the [KILT types-definitions repository](https://github.com/KILTprotocol/type-definitions/pulls)
- [x] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [x] I have documented the changes (where applicable)

### CODE REVIEW

File: traits/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+[package]
+authors = ["KILT <info@kilt.io>"]
+description = "Traits used across the KILT pallets"
+edition = "2018"
+name = "kilt-traits"
+repository = "https://github.com/KILTprotocol/mashnet-node"
+version = "0.25.0"
```
Comment (wischli) [inline traits/Cargo.toml:7]:
```suggestion
version = "0.25.1"
```

**Review by wischli — APPROVED** [review]
LGTM! Our 0.25.1 release needs to be applied to the traits pallet version though 😊

File: traits/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+/// might pay all fees and deposits that are required by the call.
+pub trait CallSources<S, P> {
+	/// The sender of the call who will pay for all deposits and fees.
+	fn sender(&self) -> S;
+
+	/// The subject of the call.
+	fn subject(&self) -> P;
+}
+
+impl<S: Clone> CallSources<S, S> for S {
```
Comment (Diiaablo95) [inline traits/src/lib.rs:33]: C: Could we also implement a trait that would not imply cloning `sender` and `subject` until it is actually needed, e.g., after verifying all conditions are met? E.g. for `AsRef<T>` or `Copy`?

File: pallets/did/src/lib.rs [hunk 3]
```diff
@@ -178,14 +179,20 @@ pub mod pallet {
...
 		/// Origin type expected by the proxied dispatchable calls.
-		type Origin: From<DidRawOrigin<DidIdentifierOf<Self>>>;
-
+		#[cfg(not(feature = "runtime-benchmarks"))]
+		type Origin: From<DidRawOrigin<DidIdentifierOf<Self>, AccountIdentifierOf<Self>>>;
 		#[cfg(feature = "runtime-benchmarks")]
 		type Origin: From<RawOrigin<DidIdentifierOf<Self>>>;
 
-		type EnsureOrigin: EnsureOrigin<Success = DidIdentifierOf<Self>, <Self as frame_system::Config>::Origin>;
+		/// The origin check for all did calls inside this pallet.
```
Comment (Diiaablo95) [inline pallets/did/src/lib.rs:188]: C:

```suggestion
		/// The origin check for all DID calls inside this pallet.
```

File: pallets/did/src/lib.rs [hunk 4]
```diff
@@ -178,14 +179,20 @@ pub mod pallet {
...
+		type EnsureOrigin: EnsureOrigin<
+			Success = <Self as Config>::OriginSuccess,
+			<Self as frame_system::Config>::Origin,
+		>;
+
+		/// The return type when the did origin check was successful.
```
Comment (Diiaablo95) [inline pallets/did/src/lib.rs:194]: C:

```suggestion
		/// The return type when the DID origin check was successful.
```

File: pallets/attestation/src/mock.rs [hunk 5]
```diff
@@ -91,6 +91,7 @@ impl frame_system::Config for Test {
 impl ctype::Config for Test {
 	type CtypeCreatorId = TestCtypeOwner;
 	type EnsureOrigin = EnsureSigned<TestCtypeOwner>;
+	type OriginSuccess = kilt_primitives::AccountId;
```
Comment (Diiaablo95) [inline pallets/attestation/src/mock.rs:94]: S:

```suggestion
	type OriginSuccess = TestCtypeOwner;
```

File: pallets/delegation/src/lib.rs [hunk 6]
```diff
@@ -129,7 +130,11 @@ pub mod pallet {
 		>;
 		type DelegationEntityId: Parameter;
 		type DelegationNodeId: Parameter + Copy + AsRef<[u8]> + Eq + PartialEq + Ord + PartialOrd;
-		type EnsureOrigin: EnsureOrigin<Success = DelegatorIdOf<Self>, <Self as frame_system::Config>::Origin>;
+		type EnsureOrigin: EnsureOrigin<
+			Success = <Self as Config>::OriginSuccess,
+			<Self as frame_system::Config>::Origin,
+		>;
+		type OriginSuccess: CallSources<<Self as frame_system::Config>::AccountId, DelegatorIdOf<Self>>;
```
Comment (Diiaablo95) [inline pallets/delegation/src/lib.rs:137]: S: should we declare and use `AccountIdOf<T>` also here?

File: pallets/delegation/src/mock.rs [hunk 7]
```diff
@@ -92,6 +92,7 @@ impl frame_system::Config for Test {
 impl ctype::Config for Test {
 	type CtypeCreatorId = TestCtypeOwner;
 	type EnsureOrigin = EnsureSigned<TestCtypeOwner>;
+	type OriginSuccess = kilt_primitives::AccountId;
```
Comment (Diiaablo95) [inline pallets/delegation/src/mock.rs:95]: S:

```suggestion
	type OriginSuccess = TestCtypeOwner;
```

File: runtimes/standalone/src/lib.rs [hunk 8]
```diff
@@ -30,7 +30,7 @@ include!(concat!(env!("OUT_DIR"), "/wasm_binary.rs"));
 #[cfg(feature = "runtime-benchmarks")]
 use frame_system::EnsureSigned;
 
-use did::DidSignature;
+use did::{DidRawOrigin, DidSignature};
```
Comment (Diiaablo95) [inline runtimes/standalone/src/lib.rs:33]: M: I guess this was automatically imported. There is now a mixed usage of `DidRawOrigin` and `did::DidRawOrigin`.

File: pallets/ctype/src/mock.rs [hunk 9]
```diff
@@ -81,6 +81,7 @@ impl frame_system::Config for Test {
 impl Config for Test {
 	type CtypeCreatorId = TestCtypeOwner;
 	type EnsureOrigin = frame_system::EnsureSigned<TestCtypeOwner>;
+	type OriginSuccess = kilt_primitives::AccountId;
```
Comment (Diiaablo95) [inline pallets/ctype/src/mock.rs:84]: S:

```suggestion
	type OriginSuccess = TestCtypeOwner;
```

**Review by Diiaablo95 — COMMENTED** [review]
I love the approach! As I did not find anything blocking this PR and it has already been accepted, I'm just commenting it. Feel free to integrate any of the MoSCoW suggestions in 👌

File: pallets/ctype/src/mock.rs (hunk 9 above)
Comment (weichweich) [reply pallets/ctype/src/mock.rs]: 2f80b2d

File: runtimes/standalone/src/lib.rs (hunk 8 above)
Comment (weichweich) [reply runtimes/standalone/src/lib.rs]: 2f80b2d

File: pallets/delegation/src/mock.rs (hunk 7 above)
Comment (weichweich) [reply pallets/delegation/src/mock.rs]: 2f80b2d

File: pallets/delegation/src/lib.rs (hunk 6 above)
Comment (weichweich) [reply pallets/delegation/src/lib.rs]: 2f80b2d

File: pallets/attestation/src/mock.rs (hunk 5 above)
Comment (weichweich) [reply pallets/attestation/src/mock.rs]: 2f80b2d

File: pallets/did/src/lib.rs (hunk 4 above)
Comment (weichweich) [reply pallets/did/src/lib.rs:194]: 2f80b2d

Comment (weichweich) [reply pallets/did/src/lib.rs:194]: 9a21625

Comment (weichweich) [reply pallets/did/src/lib.rs:194]: ah yes

**Pull request merged**